pkgs/build-support/setup-hooks/desktop-to-darwin-bundle.sh

```diff
@@ -16,24 +16,21 @@ convertDesktopFile() {
     local -r name=$(getDesktopParam "${file}" "^Name")
     local -r exec=$(getDesktopParam "${file}" "Exec")
     local -r iconName=$(getDesktopParam "${file}" "Icon")
-    local -r iconFiles=$(find "$out/share/icons/" -name "${iconName}.*" 2>/dev/null);
-    local -r pixMaps=$(find "$out/share/pixmaps/" -name "${iconName}.xpm" 2>/dev/null);
+    local -ar iconFiles=($(find "$out/share/icons/" -name "${iconName}.*" 2>/dev/null))
+    local -ar pixMaps=($(find "$out/share/pixmaps/" -name "${iconName}.xpm" 2>/dev/null))
 
     mkdir -p "$out/Applications/${name}.app/Contents/MacOS"
     mkdir -p "$out/Applications/${name}.app/Contents/Resources"
 
-    local i=0;
-    for icon in $iconFiles; do
-      ln -s "$icon" "$out/Applications/${name}.app/Contents/Resources/$i-$(basename "$icon")"
-      (( i +=1 ));
-    done
-
+    local -a convertedPixMaps
     for pixmap in $pixMaps; do
-      local newIconName="$i-$(basename "$pixmap")";
-      convert "$pixmap" "$out/Applications/${name}.app/Contents/Resources/${newIconName%.xpm}.png"
-      (( i +=1 ));
+        local newIconName=$(mktemp --suffix=.png);
+        convertedPixMaps+=($newIconName)
+        convert "$pixmap" "$newIconName"
     done
 
+    png2icns "$out/Applications/${name}.app/Contents/Resources/${name}.icns" "${iconFiles[@]}" "${convertedPixMaps[@]}"
```

Comment:
I tried something similar; unfortunately, you hit this problem:

```
Bad dimensions: PNG file '/nix/store/7g946nsnmslg4vqsayjskj37fmy3zcqz-element-desktop-1.10.1/share/icons/hicolor/96x96/apps/element.png' is 96x96
```

It turns out the `.icns` only allows specific resolutions of png files.  Not only that, for some resolutions, only specific bit depths are permitted. See the big switch statement [here](https://github.com/kornelski/libicns/blob/1137b6f67eed87fe7a31c7de5325686422bc4568/src/icns_utils.c#L536l).  At that point I gave up and just copied the PNG files into the `.app` file!

Reply:
I’d agree if we had a way to specify which type to use, but the version packaged in nixpkgs doesn’t allow that. I assume it will usually do the right thing. That’s a fork though. The upstream has a clone of `iconutil` that supports `iconset`s, but it hasn’t had a release, and `autoreconf` fails to generate a working `configure` script. Unfortunately, the [master branch][1] of the original upstream is the only way to support retina icons using this tool. 😑

[1]: https://sourceforge.net/p/icns/code/ci/master/tree/